Lib: SMF: fix stuck handled bit when transitioning

glenn-andrews opened the pull request:
If both `smf_set_state()` and `smf_set_handled()` are called in a run action, the `internal->handled` bit would not be reset.

This could cause an issue on the next run action by not propagating events to parents.

glenn-andrews committed:
Lib: SMF: fix stuck handled bit when transitioning

If both `smf_set_state()` and `smf_set_handled()` are called in
a run action, the `internal->handled` bit would not be reset.

This could cause an issue on the next run action by not
propagating events to parents.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

keith-zephyr reviewed (dismissed):
Looks good to me.  Is it possible to add or modify a test so that it makes calls to both `smf_set_state` and `smf_set_handled` from a run action?

glenn-andrews commented:
> Looks good to me. Is it possible to add or modify a test so that it makes calls to both `smf_set_state` and `smf_set_handled` from a run action?

Currently there's a failed test, so something was relying on the bad functionality. I'll dig into it this weekend.

glenn-andrews commented:
Found the issue. Initializing tests was not resetting all internal state bits.

I'll look at improving the tests to check smf_set_handled() works.

glenn-andrews committed:
Lib: SMF: Refactor test_lib_self_transition_smf.c

Make the code less dependent on
magic #defines by using a single enum for test states.

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews committed:
Lib: SMF: Add test for stuck handled bit

Update test_lib_self_transition_smf.c to detect the bug that was in
smf_set_handled().

Also delete unneeded test state STATE_D_ENTRY

Signed-off-by: Glenn Andrews <glenn.andrews.42@gmail.com>

glenn-andrews commented:
OK, confirmed the updated test finds the old bug.

keith-zephyr commented on lib/smf/smf.c:
```diff
@@ -229,7 +221,8 @@ void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
 	internal->terminate = false;
-	ctx->current = init_state;
+	internal->handled = false;
+	internal->new_state = false;	ctx->current = init_state;
```
Formatting quirk.  These statements should be on separate lines.

glenn-andrews replied:
Oh no :( I'm not sure I'm enough of a Git to know how to amend an earlier commit. Can I fix it by amending the last commit?

keith-zephyr replied:
FYI, my workflow is to use "fixup" commits to make changes to a multi-commit stack.

1. Make the changes to the unstaged file
2. Run `git commit -a --fixup <commit_hash>`.  For example `git commit -a --fixup HEAD~2` to fixup the 3rd commit from the top.  Or use the direct hash shown by `git log --oneline`
3. Repeat steps 1&2 for additional changes that need to target different commits.
4. Run `git rebase -i --autosquash`

keith-zephyr commented on tests/lib/smf/src/test_lib_self_transition_smf.c:
```diff
@@ -222,6 +220,12 @@ static void parent_ab_run(void *obj)
+	 * There was a bug that did not reset the handled bit if both were called,
+	 * so check it's still fixed:
+	 */
+	smf_set_handled(SMF_CTX(obj));
```
You're explicitly violating the API contract by calling `smf_set_handled() after `smf_set_state()`.  

If you swap these calls (`smf_set_handled()` then `smf_set_state()`), does the test still detect the problem?

glenn-andrews replied:
To be honest the API contract should state `smf_set_handled()` is never called in the same code path as `smf_set_state()` or `smf_set_terminate()`, before or after.

Should we error out if people do, or should we accept it as this fix does?

keith-zephyr replied:
The current contract states that it's not required to call `smf_set_handled()` if calling `smf_set_state()`.

You could potentially check if `smf_set_handled()` is called before `smf_set_state()`, but I don't think you could check for the opposite order.

So it's probably best to leave as is, but swap the calling order in the test, as we don't want to have any examples in the tree showing a state doing work after calling `smf_set_state()`.

glenn-andrews replied:
Will do.

glenn-andrews replied:
Confirmed the test fails if `smf_set_handled()` is called before `smf_set_state()` on main.

I'm teaching using SMF, basically telling them "If you handle the event and don't want it propagated to parents, call `smf_set_handled()` except after `smf_set_state()` or `smf_set_terminate()`", and some didn't hear the last bit.

The _correct_ way to fix this would be to put in a return code to state handler functions that returns handled/propagate to parent/ transition occurred/terminate to force people to choose what the state machine does after the handler has run. But doing so would break every existing SMF usage out here, so we're stuck with the `smf_set_handled()` kluge, which either gets forgotten, or put places it's not needed.

keith-zephyr commented:
@sambhurst is OOO at the moment, so reassigned to myself.
